samza-core/src/main/java/org/apache/samza/clustermanager/ResourceRequestState.java

```diff
@@ -266,18 +268,34 @@ private void clearState() {
   /**
    * Retrieves, but does not remove, the first allocated resource on the specified host.
    *
-   * @param host  the host for which a resource is needed.
-   * @return      the first {@link SamzaResource} allocated for the specified host or {@code null} if there isn't one.
+   * @param host the host for which a resource is needed.
+   * @return a {@link SamzaResource} allocated for the specified host or {@code null} if there isn't one.
    */
-
   public SamzaResource peekResource(String host) {
     synchronized (lock) {
-      List<SamzaResource> resourcesOnTheHost = this.allocatedResources.get(host);
+      List<SamzaResource> resourcesOnPreferredHostBuffer = this.allocatedResources.get(host);
+      List<SamzaResource> resourcesOnAnyHostBuffer = this.allocatedResources.get(ANY_HOST);
 
-      if (resourcesOnTheHost == null || resourcesOnTheHost.isEmpty()) {
+      // First search for the preferred host buffers
+      if (resourcesOnPreferredHostBuffer != null && !resourcesOnPreferredHostBuffer.isEmpty()) {
+        SamzaResource resource = resourcesOnPreferredHostBuffer.get(0);
+        log.info("Returning a buffered resource: {} for {} from preferred-host buffer.", resource.getResourceID(), host);
+        return resource;
+      } else if (resourcesOnAnyHostBuffer != null && !resourcesOnAnyHostBuffer.isEmpty()) {
+        // If preferred host buffers are empty, scan the ANY_HOST buffer
+        log.debug("No resources on preferred-host buffer. Scanning ANY_HOST buffer");
+        SamzaResource resource = resourcesOnAnyHostBuffer.stream()
+            .filter(resrc -> resrc.getHost().equals(host))
+            .findAny().orElse(null);
+        if (resource != null) {
+          log.info("Returning a buffered resource: {} for {} from ANY_HOST buffer.", resource.getResourceID(), host);
+        }
+        return resource;
+      } else {
+        // If both preferred host and ANY_HOST buffers are empty, return null
+        log.debug("Cannot find any resource in the ANY_HOST buffer for {} because it is empty", host);
```

Comment:
Update message to say that we could not find a resource in either buffer because _both_ buffers are empty. Maybe log info, or even warn?